### PR TITLE
tools/filegone: add filegone to trace why file gone

### PR DESCRIPTION
**Pull request by curu**
file gone/vanished either been deleted or renamed

### CODE REVIEW

File: tools/filevanish.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 12, 0)
+// trace file deletion and output details
+int trace_unlink(struct pt_regs *ctx, struct inode *dir, struct dentry *dentry)
+#else
+int trace_unlink(struct pt_regs *ctx, struct user_namespace *ns, struct inode *dir, struct dentry *dentry)
+#endif
+{
+    struct data_t data = {};
+    u32 pid = bpf_get_current_pid_tgid();
```
Comment (chenhengqi) [inline tools/filevanish.py:59]: `pid` should be `bpf_get_current_pid_tgid() >> 32`.

File: tools/filevanish.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+    FILTER
+
+    struct qstr d_name = dentry->d_name;
+    if (d_name.len == 0)
+        return 0;
+
+    if (bpf_get_current_comm(&data.comm, sizeof(data.comm)) == 0) {
```
Comment (chenhengqi) [inline tools/filevanish.py:67]: No need to check the return value of `bpf_get_current_comm()`, even it fails, we should capture this event.

File: tools/filevanish.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+struct inode *new_dir, struct dentry *new_dentry)
+{
+#else
+int trace_rename(struct pt_regs *ctx, struct renamedata *rd)
+{
+    struct dentry *old_dentry = rd->old_dentry;
+    struct dentry *new_dentry = rd->new_dentry;
+#endif
+    struct data_t data = {};
+    u32 pid = bpf_get_current_pid_tgid();
```
Comment (chenhengqi) [inline tools/filevanish.py:89]: ditto

File: tools/filevanish.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        exit()
+
+# initialize BPF
+b = BPF(text=bpf_text)
+b.attach_kprobe(event="vfs_unlink", fn_name="trace_unlink")
+b.attach_kprobe(event="vfs_rmdir", fn_name="trace_unlink")
+b.attach_kprobe(event="vfs_rename", fn_name="trace_rename")
+
+# header
+print("%-8s %-7s %-16s %7s %s" % ("TIME", "PID", "COMM", "ACTION", "FILE"))
```
Comment (chenhengqi) [inline tools/filevanish.py:135]: `%6s` for `ACTION`.

File: tools/filevanish.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    u8 action;
+    char comm[TASK_COMM_LEN];
+    char fname[DNAME_INLINE_LEN];
+    char fname2[DNAME_INLINE_LEN];
+};
+
+BPF_HASH(birth, struct dentry *);
+BPF_PERF_OUTPUT(events);
+
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 12, 0)
```
Comment (chenhengqi) [inline tools/filevanish.py:51]: We could use BTF info to avoid this version check, see #4339.

**Comment by brendangregg** [comment]
Thanks, some comments:

- I like it. Seems like a useful debugging tool, and a companion to filelife.
- Maybe filegone is a better name? Short is good.
- Needs a man page, and a link in README.md.
- I've come across a temporary file mechanism in libc/syscalls before, I forget the name, but seemed to be a special way to create temporary file objects that didn't go via the typical paths. Doesn't have to go in this version, but someday it would be good to trace it and add it to filelife and filegone.

File: tools/filevanish.py (hunk 5 above)
Comment (curu) [reply tools/filevanish.py]: maybe change it later.
currently we need to keep compatibility with bcc versions below v0.23 which has no 'kernel_struct_has_field'

Comment (Rtoax) [reply tools/filevanish.py]: Hi, can we do this https://github.com/iovisor/bcc/pull/4346 ?

Comment (curu) [reply tools/filevanish.py]: > Hi, can we do this #4346 ?
RHEL/CentOS 7 comes with bcc version 0.10, which has no **kernel_struct_has_field**,   so the same script woud not run without modification.

so maybe we should drop support for that old version and use a legacy branch to do that?

**Pull request merged**